server/routerlicious/packages/services-telemetry/src/lumberjack.ts

```diff
@@ -0,0 +1,111 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "path";
+import { LumberEventName } from "./lumberEventNames";
+import { Lumber } from "./lumber";
+import { LogLevel, LumberType, ITelemetryMetadata, ILumberjackEngine } from "./resources";
+
+// Lumberjack is a telemetry manager class that allows the collection of metrics and logs
+// throughout the service. A list of ILumberjackEngine must be provided to Lumberjack
+// by calling setupEngines() before Lumberjack can be used.
+export class Lumberjack {
+    protected static _instance: Lumberjack | undefined = undefined;
+    private readonly engineList: ILumberjackEngine[];
+    private isSetupCompleted: boolean;
+
+    protected constructor() {
+        this.engineList = [];
+        this.isSetupCompleted = false;
+    }
+
+    public static get instance() {
+        if (!Lumberjack._instance) {
+            Lumberjack._instance = new Lumberjack();
+        }
+
+        return Lumberjack._instance;
+    }
+
+    public setupEngines(engines: ILumberjackEngine[]) {
+        if (this.isSetupCompleted) {
+            throw new Error("This Lumberjack was already setup with a list of engines.");
+        }
+
+        this.engineList.push(...engines);
+        this.isSetupCompleted = true;
+    }
+
+    public newLumberMetric<T extends string = LumberEventName>(eventName: T) {
+        this.throwOnEmptyEngineList();
+        return new Lumber<T>(eventName, LumberType.Metric, this.engineList);
+    }
+
+    public log(
+        message: string,
+        metadata: ITelemetryMetadata,
+        level: LogLevel,
+        properties?: Map<string, any> | Record<string, any> | undefined,
+        statusCode?: number | string | undefined,
+        exception?: Error | undefined,
+    ) {
+        this.throwOnEmptyEngineList();
+        const lumber = new Lumber<string>(this.getLogCallerInfo(), LumberType.Log, this.engineList);
+
+        if (properties) {
+            if (properties instanceof Map) {
+                properties.forEach((value: any, key: string) => { lumber.addProperty(key, value); });
+            } else {
+                for (const [key, value] of Object.entries(properties)) {
```

Comment:
could we do something like the following to be consistent with the above for maps?
```
Object.entries(properties).forEach([key,value] => lumber.addProperty(key,value))
```